python/cudf/cudf/utils/utils.py

```diff
@@ -136,43 +116,56 @@ def buffers_from_pyarrow(pa_arr, dtype=None):
         - cudf.Buffer --> data
         - cudf.Buffer --> string characters
     """
-    from cudf.core.buffer import Buffer
-    from cudf.utils.cudautils import copy_array
-
     buffers = pa_arr.buffers()
 
     if pa_arr.null_count:
-        mask_dev_array = make_mask(len(pa_arr))
-        arrow_dev_array = rmm.to_device(np.asarray(buffers[0]).view("int8"))
-        copy_array(arrow_dev_array, mask_dev_array)
-        pamask = Buffer(mask_dev_array)
+        mask_size = bitmask_allocation_size_bytes(len(pa_arr))
+        pamask = pyarrow_buffer_to_cudf_buffer(buffers[0], mask_size=mask_size)
     else:
         pamask = None
 
     offset = pa_arr.offset
     size = len(pa_arr)
 
-    if dtype:
-        data_dtype = dtype
-    elif isinstance(pa_arr, pa.StringArray):
-        data_dtype = np.int32
-    else:
-        if isinstance(pa_arr, pa.DictionaryArray):
-            data_dtype = pa_arr.indices.type.to_pandas_dtype()
-        else:
-            data_dtype = pa_arr.type.to_pandas_dtype()
-
     if buffers[1]:
-        padata = Buffer(np.asarray(buffers[1]).view(data_dtype))
+        padata = pyarrow_buffer_to_cudf_buffer(buffers[1])
     else:
         padata = Buffer.empty(0)
 
     pastrs = None
     if isinstance(pa_arr, pa.StringArray):
-        pastrs = Buffer(np.asarray(buffers[2]).view(np.int8))
+        pastrs = pyarrow_buffer_to_cudf_buffer(buffers[2])
     return (size, offset, pamask, padata, pastrs)
 
 
+def pyarrow_buffer_to_cudf_buffer(arrow_buf, mask_size=0):
+    """
+    Given a PyArrow Buffer backed by either host or device memory, convert it
+    to a cuDF Buffer
+    """
+    # Try creating a PyArrow CudaBuffer from the PyArrow Buffer object, it
+    # fails with an ArrowTypeError if it's a host based Buffer so we catch and
```

Comment:
Is this the only case in which we can get a `ArrowTypeErrror`?

Reply:
Yea, the objects passed to this function are guaranteed to be Arrow Buffers in how we're using it currently. I could add an `isinstance` check and explicitly error if we wanted.

Reply:
Yeah that'd be safer

Reply:
done